src/Encoder/JsonApi.php

```diff
@@ -0,0 +1,201 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Encoder;
+
+use App\Normalizer\JsonApiDTO;
+use App\Service\EntityManagerLookup;
+use Symfony\Component\DependencyInjection\ContainerAwareTrait;
+use Symfony\Component\Serializer\Encoder\DecoderInterface;
+use Symfony\Component\Serializer\Encoder\EncoderInterface;
+use Exception;
+
+class JsonApi implements EncoderInterface, DecoderInterface
+{
+    use ContainerAwareTrait;
+
+    protected const FORMAT = 'json-api';
+
+    /**
+     * @var EntityManagerLookup
+     */
+    protected $entityManagerLookup;
+
+    /**
+     * @var JsonApiDTO
+     */
+    protected $normalizer;
+
+    public function __construct(EntityManagerLookup $entityManagerLookup, JsonApiDTO $normalizer)
+    {
+
+        $this->entityManagerLookup = $entityManagerLookup;
+        $this->normalizer = $normalizer;
+    }
+
+    public function decode(string $data, string $format, array $context = [])
+    {
+        // TODO: Implement decode() method.
+    }
+
+    public function supportsDecoding(string $format)
+    {
+        // TODO: Implement supportsDecoding() method.
+    }
+
+    public function encode($data, string $format, array $context = [])
+    {
+        $shaped = $this->shapeData($data, $this->extractSideLoadFields($context));
+
+        if (array_key_exists('singleItem', $context) && $context['singleItem']) {
+            $data = $shaped['data'];
+            $item = $data[0];
+            $shaped['data'] = $item;
+        }
+
+        return json_encode($shaped);
+    }
+
+    protected function shapeData(array $data, array $sideLoadFields): array
```

Comment:
suggesting to refactor the the `shapeData`/`shapeItem` functionality out into its own DataShaper service.

Reply:
Good call. Will do.